### DISCUSSION TITLE
How do I print only the shape with jax.debug.print

### DISCUSSION THREAD

**Discussion by Logon27** [jax-ml/jax#17313] [Q&A] [answered]
Hello, I am using jax.debug.print in order to print the shape of an array. However when I actually call `result.shape` it produces...
`(array(128, dtype=int32), array(1024, dtype=int32))`
Which is a tuple of arrays with the dtype included. And I do not consider dtypes to be shape information. How would I go about getting it to only print...
`(128, 1024)`
Any help is appreciated. I have tried a lot of different things, but to no avail.

**Comment by ASEM000** [comment]
You can use the following code:

```python

import jax

import jax.tree_util as jtu


@jtu.register_pytree_node_class
class StaticShape:
    def __init__(self, value: tuple[int, ...]):
        self.value = value

    def tree_flatten(self):
        return (), self.value

    @classmethod
    def tree_unflatten(cls, aux_data, _):
        self = object.__new__(cls)
        self.value = aux_data
        return self

    def __repr__(self):
        return f"{self.value!r}"

    def __hash__(self):
        return hash(self.value)

    def __eq__(self, other):
        return (self.value == other.value) if isinstance(other, StaticShape) else False


@jax.jit
def f(x):
    jax.debug.print("shape={x}", x=StaticShape(x.shape))
    return x


f(jax.numpy.ones([4, 4]))
# shape=(4, 4)
# Array([[1., 1., 1., 1.],
#        [1., 1., 1., 1.],
#        [1., 1., 1., 1.],
#        [1., 1., 1., 1.]], dtype=float32)
```

I suspect this issue is due to `jax.jit` changing your tuple of ints  to a tuple of arrays.

**Reply by Logon27** [reply]
This should work fine for my use case. A bit unfortunate that I need a custom class to achieve proper shape printing, but it is what it is. Thank you very much for the help.

**Reply by ASEM000** [reply]
jax team will introduce `jtu.register_static` to simplify this process by next release.

**Comment by jakevdp** [comment, marked as answer]
`jax.debug.print` is designed for printing dynamic values. Array shapes are always static values, so the better tool would be a standard Python print:
```python
# print dynamic array contents with debug print:
jax.debug.print("{}", results)

# print static array attributes with Python print:
print(results.shape)
print(results.dtype)
```

**Reply by Logon27** [reply]
Edit: I see what you are saying now about the sizes of the arrays being static. Are there any side effects to me printing shape information inside a jit'd function using a standard python print?

**Reply by jakevdp** [reply]
It is always safe to print static values like shapes, even within a jit-compiled function (give it a try!). You only need `jax.debug.print` for dynamic values like the contents of arrays.

**Reply by Logon27** [reply]
Ah I see. I did not understand the differentiation of the use cases between printing static and dynamic values when reading the jax.debug.print or the sharp bits docs. The jax.debug.print documentation does not use the verbage 'static' or 'dynamic' at all. So that was just a misunderstanding on my part really. I will try this later today. Thanks for the help!

**Reply by jakevdp** [reply]
Thanks - that's useful feedback. I've proposed some changes in #17331 that will hopefully help

**Reply by ASEM000** [reply]
As a follow-up, one difference between the two approaches is how many prints each does: 

Consider the following:

```python

@jax.jit
def f(x):
    jax.debug.print("shape={x}", x=StaticShape(x.shape))
    return x

print("-----jax debug print-----")
for i in range(5):
    print(f"i={i}")
    f(jax.numpy.ones([4, 4]))
print("-----jax debug print-----")
for i in range(5):
    print(f"i={i}")
    f(jax.numpy.ones([4, 4]))
# -----jax debug print-----
# i=0
# shape=(4, 4)
# i=1
# shape=(4, 4)
# i=2
# shape=(4, 4)
# i=3
# shape=(4, 4)
# i=4
# shape=(4, 4)
```


```python

@jax.jit
def f(x):
    print("shape={x}".format(x=x.shape))
    return x

print("-----python debug print-----")
for i in range(5):
    print(f"i={i}")
    f(jax.numpy.ones([4, 4]))
# -----python debug print-----
# i=0
# shape=(4, 4)
# i=1
# i=2
# i=3
# i=4
```

Maybe it is worth mentioning the behavior of `print` inside a jtted function in the context of `jax.debug.print` docs ?

**Reply by jakevdp** [reply]
That's the difference between trace-time and runtime: it's pretty well covered here I think? https://jax.readthedocs.io/en/latest/notebooks/thinking_in_jax.html

**Reply by ASEM000** [reply]
> it's pretty well covered here I think?

Yes, 

> in the context of jax.debug.print docs

My suggestion is that, if to suggest `print` inside the docs, maybe point out this behavior also _within_ the `jax.debug.print` docs as a direct comparison between the approaches.

**Reply by Logon27** [reply]
@ASEM000 Ah I see what you mean about the difference in the number of prints. I may still want your solution then. I guess I will have to test and see if I think a single print is sufficient in the context of my own program. Because the behavior might be odd if I mix and match both regular prints and jax.debug.print. Because it would display the regular print on the first iteration, but not on subsequent iterations while jax.debug.print would. Which might be confusing. A pretty niche issue. I think both your answers are correct. It just depends on which behavior is sufficient for what the individual developer is trying to do.

**Reply by jakevdp** [reply]
Thanks - makes sense.

Note that if you need the static quantity to print at every loop iteration, you could always do something like this:
```python
jax.debug.print(f"{arr.shape}")
```
Since it's a static quantity, you can put it in the static string argument that you pass to `debug.print`.

Defining and registering a pytree just to print a shape seems like a little too heavy of a solution here, if I'm being honest.

**Reply by ASEM000** [reply]
> Defining and registering a pytree just to print a shape seems like a little too heavy of a solution here, if I'm being honest.

I think you are absolutely right, I'm guilty of overdosing on the use of pytrees :D
